contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2803,6 +3010,714 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+		 */
+		current_permission = old_permission & ~new_permission;
+	}
+
+	if (current_permission == 0)
+	{
+		remove_entry_from_bbf_schema_perms(schema_name, object_name, grantee, object_type);
+		return;
+	}
+
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+							RowExclusiveLock);
+
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ,
+				CStringGetTextDatum(schema_name));
+	ScanKeyEntryInitialize(&scanKey[2], 0,
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ,
+				CStringGetTextDatum(object_name));
+	ScanKeyInit(&scanKey[3],
+				Anum_bbf_schema_perms_permission,
+				BTEqualStrategyNumber, F_INT4EQ,
+				Int32GetDatum(old_permission));
+	ScanKeyEntryInitialize(&scanKey[4], 0,
+				Anum_bbf_schema_perms_grantee,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ,
+				CStringGetTextDatum(grantee));
+
+	scan = systable_beginscan(bbf_schema_rel,
+				get_bbf_schema_perms_idx_oid(),
+				false, NULL, 5, scanKey);
+
+	tuple_bbf_schema = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple_bbf_schema))
+	{
+		bbf_schema_dsc = RelationGetDescr(bbf_schema_rel);
+		/* Build a tuple to insert */
+		MemSet(new_record_bbf_schema, 0, sizeof(new_record_bbf_schema));
+		MemSet(new_record_nulls_bbf_schema, false, sizeof(new_record_nulls_bbf_schema));
+		MemSet(new_record_repl_bbf_schema, false, sizeof(new_record_repl_bbf_schema));
+
+		new_record_bbf_schema[Anum_bbf_schema_perms_permission - 1] = Int32GetDatum(current_permission);
+		new_record_repl_bbf_schema[Anum_bbf_schema_perms_permission - 1] = true;
+
+		new_tuple = heap_modify_tuple(tuple_bbf_schema,
+									bbf_schema_dsc,
+									new_record_bbf_schema,
+									new_record_nulls_bbf_schema,
+									new_record_repl_bbf_schema);
+
+		CatalogTupleUpdate(bbf_schema_rel, &new_tuple->t_self, new_tuple);
+		heap_freetuple(new_tuple);
+	}
+
+	systable_endscan(scan);
+	table_close(bbf_schema_rel, RowExclusiveLock);
+}
+
+/*
+ * Checks if a particular privilege exists in catalog BABELFISH_SCHEMA_PERMISSIONS.
+ */
+bool
+privilege_exists_in_bbf_schema_permissions(const char *schema_name,
+							const char *object_name,
+							const char *grantee,
+							bool ignore_grantee)
+{
+	Relation	bbf_schema_rel;
+	HeapTuple	tuple_bbf_schema;
+	SysScanDesc	scan;
+	bool	catalog_entry_exists = false;
+	int16	dbid = get_cur_db_id();
+
+	/* Immediately return false, if SCHEMA name is NULL or it's a shared schema. */
+	if (schema_name == NULL || is_shared_schema(schema_name))
+		return false;
+
+	if (!ignore_grantee)
```

Comment:
No, there won't be any such case. 
ignore_grantee = true only where grantee is NULL and we have to avoid look up on grantee since it is NULL and grantee_name is a NOT NULL value in the catalog.

Reply:
Then, why do we need this extra boolean argument? We should just check whether grantee == NULL.